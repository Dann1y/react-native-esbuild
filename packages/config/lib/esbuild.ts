import path from 'node:path';
import deepmerge from 'deepmerge';
import type { BuildOptions } from 'esbuild';
import type { EsbuildPresetOptions } from './types';
import { ASSET_EXTENSIONS, BANNER_VARS, SOURCE_EXTENSIONS } from './shares';

export function getEsbuildOptions(
  options: EsbuildPresetOptions,
  customEsbuildOptions?: Partial<BuildOptions>,
): BuildOptions {
  const { entryPoint, outfile, assetsDir, platform, dev, minify } = options;

  const platforms = [platform, 'native', 'react-native'];
  const extensions = SOURCE_EXTENSIONS.concat(ASSET_EXTENSIONS);

  const getReactNativePolyfills = (): string[] => {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const getPolyfills = require(require.resolve(
      'react-native/rn-get-polyfills',
      {
        paths: Array.from(
          new Set([
            // add current workspace directory to module resolution path.
            // improve for workspace projects (eg. monorepo)
            path.join(process.cwd(), 'node_modules'),
            ...(require.main?.paths ?? []),
          ]),
        ).filter(Boolean),
      },
    )) as () => string[];

    return getPolyfills();
  };

  const baseOptions: BuildOptions = {
    /**
     * mainFields
     * @see {@link https://github.com/facebook/metro/blob/0.72.x/docs/Configuration.md#resolvermainfields}
     */
    mainFields: ['react-native', 'browser', 'main', 'module'],
    entryPoints: [entryPoint],
    outfile,
    assetNames: `${assetsDir}/[name]-[hash]`,
    sourceRoot: process.cwd(),
    resolveExtensions: platforms
      .map((platform) => extensions.map((ext) => `.${platform}${ext}`))
      .concat(extensions)
      .flat(),
    define: {
      __DEV__: JSON.stringify(dev),
      global: 'window',
      'process.env.NODE_ENV': JSON.stringify(
        dev ? 'development' : 'production',
      ),
    },
    loader: Object.fromEntries(ASSET_EXTENSIONS.map((ext) => [ext, 'file'])),
    legalComments: 'none',
    banner: {
      js: `var ${BANNER_VARS.join(',')};`,
    },
    inject: [
      'react-native/Libraries/Core/InitializeCore',
      ...getReactNativePolyfills(),
    ],
    target: 'es6',
    supported: {
      /**
       * to avoid block scope bug on hermes engine.
       * in `__copyProps`, generated by esbuild
       *
       * @see {@link https://github.com/evanw/esbuild/blob/v0.18.17/internal/runtime/runtime.go#L192}
       */
      'for-of': false,
    },
    format: 'iife',
    bundle: true,
    sourcemap: true,
    minify,
    logLevel: 'silent',
  };

  return deepmerge(baseOptions, customEsbuildOptions ?? {});
}
