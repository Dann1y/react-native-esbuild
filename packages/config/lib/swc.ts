import deepmerge from 'deepmerge';
import type { EsParserConfig, Options, TsParserConfig } from '@swc/core';
import type { SwcPresetOptions } from './types';

export function getSwcOptions(
  options: SwcPresetOptions,
  customSwcOptions?: Partial<Options>,
): Options {
  const isTS = /\.tsx?$/.test(options.filename);
  const parseOption = isTS
    ? ({
        syntax: 'typescript',
        tsx: true,
        dynamicImport: true,
      } as TsParserConfig)
    : ({
        syntax: 'ecmascript',
        jsx: true,
        exportDefaultFrom: true,
      } as EsParserConfig);

  const baseOptions: Options = {
    isModule: true,
    sourceMaps: false,
    jsc: {
      parser: parseOption,
      target: 'es5',
      loose: true,
      externalHelpers: true,
      keepClassNames: true,
    },
  };

  return customSwcOptions
    ? deepmerge(baseOptions, customSwcOptions)
    : baseOptions;
}
