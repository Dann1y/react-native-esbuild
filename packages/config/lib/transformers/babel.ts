import deepmerge from 'deepmerge';
import type { TransformOptions } from '@babel/core';

export function getBabelOptions(
  customBabelOptions?: Partial<TransformOptions>,
): TransformOptions {
  const baseOptions: TransformOptions = {
    minified: false,
    compact: false,
    sourceMaps: false,
  };

  return customBabelOptions
    ? deepmerge(baseOptions, customBabelOptions)
    : baseOptions;
}
