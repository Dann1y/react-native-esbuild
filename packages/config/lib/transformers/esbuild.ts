import path from 'node:path';
import deepmerge from 'deepmerge';
import type { BuildOptions } from 'esbuild';
import type { BundleOptions } from '../types';

export const getEsbuildOptions = (
  bundleOptions: BundleOptions,
  customEsbuildOptions?: Partial<BuildOptions>,
): BuildOptions => {
  const { entry, outfile, minify, metafile } = bundleOptions;

  const baseOptions: BuildOptions = {
    entryPoints: [entry],
    outfile,
    sourceRoot: path.dirname(entry),
    legalComments: 'none',
    target: 'es6',
    supported: {
      /**
       * to avoid block scope bug on hermes engine.
       * in `__copyProps`, generated by esbuild
       *
       * @see {@link https://github.com/evanw/esbuild/blob/v0.18.17/internal/runtime/runtime.go#L192}
       */
      'for-of': false,
    },
    format: 'esm',
    bundle: true,
    sourcemap: true,
    minify,
    metafile,
    logLevel: 'silent',
  };

  return deepmerge(baseOptions, customEsbuildOptions ?? {});
};
