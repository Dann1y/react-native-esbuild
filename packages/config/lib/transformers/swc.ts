import deepmerge from 'deepmerge';
import type { EsParserConfig, Options, TsParserConfig } from '@swc/core';
import type { SwcPresetOptions } from '../types';

export function getSwcOptions(
  options: SwcPresetOptions,
  customSwcOptions?: Partial<Options>,
): Options {
  const isTS = /\.tsx?$/.test(options.filename);
  const parseOption = isTS
    ? ({
        syntax: 'typescript',
        tsx: true,
        dynamicImport: true,
      } as TsParserConfig)
    : ({
        syntax: 'ecmascript',
        jsx: true,
        exportDefaultFrom: true,
      } as EsParserConfig);

  const baseOptions: Options = {
    minify: false,
    sourceMaps: false,
    isModule: true,
    inputSourceMap: false,
    inlineSourcesContent: false,
    root: options.root,
    jsc: {
      parser: parseOption,
      target: 'es5',
      loose: false,
      externalHelpers: true,
      keepClassNames: true,
    },
  };

  return customSwcOptions
    ? deepmerge(baseOptions, customSwcOptions)
    : baseOptions;
}
