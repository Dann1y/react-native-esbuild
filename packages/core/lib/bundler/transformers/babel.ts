import {
  loadOptions,
  transform,
  type TransformOptions,
  type BabelFileResult,
} from '@babel/core';
import { getBabelOptions } from '@react-native-esbuild/config';
import { promisify } from '../../helpers';
import type { Transformer } from '../../types';

export const transformWithBabel: Transformer<TransformOptions> = async (
  code: string,
  context,
  customOptions,
) => {
  const options = loadOptions({
    ...getBabelOptions(context.root, customOptions),
    filename: context.args.path,
    caller: {
      name: '@react-native-esbuild/plugins',
    },
  });

  if (!options) {
    throw new Error('cannot load babel options');
  }

  const { code: transformedCode } = await promisify<BabelFileResult>(
    (handler) => transform(code, options, handler),
  );

  if (typeof transformedCode !== 'string') {
    throw new Error('babel transformed source is empty');
  }

  return transformedCode;
};
