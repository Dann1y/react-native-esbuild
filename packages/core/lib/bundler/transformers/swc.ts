import { transform, type Options } from '@swc/core';
import { getSwcOptions } from '@react-native-esbuild/config';
import type { Transformer } from '../../types';

export const transformWithSwc: Transformer<Options> = async (
  code,
  context,
  customOptions,
) => {
  const options = getSwcOptions(
    {
      filename: context.args.path,
      root: context.root,
    },
    customOptions,
  );
  const { code: transformedCode } = await transform(code, options);

  if (typeof transformedCode !== 'string') {
    throw new Error('swc transformed source is empty');
  }

  return transformedCode;
};
