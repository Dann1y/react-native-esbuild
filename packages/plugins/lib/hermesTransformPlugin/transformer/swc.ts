import { transform } from '@swc/core';
import type { OnLoadArgs } from 'esbuild';
import { getSwcOptions } from '@react-native-esbuild/config';

export const transformWithSwc = async (
  source: string,
  args: OnLoadArgs,
): Promise<string> => {
  const options = getSwcOptions({ filename: args.path });
  const { code } = await transform(source, options);

  if (typeof code !== 'string') {
    throw new Error('swc transformed source is empty');
  }

  return code;
};
