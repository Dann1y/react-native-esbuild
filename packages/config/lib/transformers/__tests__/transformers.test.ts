import { getEsbuildOptions } from '../esbuild';
import { getSwcOptions } from '../swc';
import { getBabelOptions } from '../babel';

describe('getEsbuildOptions', () => {
  it('should match snapshot', () => {
    expect(
      getEsbuildOptions({
        entryPoint: 'index.js',
        assetsDir: 'assets',
        dev: true,
        minify: false,
        outfile: 'main.bundlejs',
        platform: 'android',
      }),
    ).toMatchSnapshot();
  });
});

describe('getSwcOptions', () => {
  it('should match snapshot', () => {
    expect(getSwcOptions({ filename: 'file.js' })).toMatchSnapshot();
  });

  describe('when filename ends with `.js`', () => {
    it('should set parse syntax to `ecmascript`', () => {
      const options = getSwcOptions({ filename: 'file.js' });
      expect(options.jsc?.parser?.syntax).toEqual('ecmascript');
    });
  });

  describe('when filename ends with `.jsx`', () => {
    it('should set parse syntax to `ecmascript`', () => {
      const options = getSwcOptions({ filename: 'file.jsx' });
      expect(options.jsc?.parser?.syntax).toEqual('ecmascript');
    });
  });

  describe('when filename ends with `.ts`', () => {
    it('should set parse syntax to `typescript`', () => {
      const options = getSwcOptions({ filename: 'file.ts' });
      expect(options.jsc?.parser?.syntax).toEqual('typescript');
    });
  });

  describe('when filename ends with `.tsx`', () => {
    it('should set parse syntax to `typescript`', () => {
      const options = getSwcOptions({ filename: 'file.tsx' });
      expect(options.jsc?.parser?.syntax).toEqual('typescript');
    });
  });

  describe('when customSwcOptions present', () => {
    it('should override default options', () => {
      const overrideOptions = {
        isModule: false,
        jsc: {
          parser: undefined,
          loose: false,
        },
      };
      const options = getSwcOptions({ filename: 'file.js' }, overrideOptions);

      expect(options).toMatchObject(
        expect.objectContaining({
          ...overrideOptions,
          jsc: expect.objectContaining(overrideOptions.jsc),
        }),
      );
    });
  });
});

describe('getBabelOptions', () => {
  it('should match snapshot', () => {
    expect(getBabelOptions()).toMatchSnapshot();
  });

  describe('when customBabelOptions present', () => {
    it('should override default options', () => {
      const overrideOptions = {
        minified: true,
        compact: true,
        sourceMaps: true,
      };
      const options = getBabelOptions(overrideOptions);

      expect(options).toMatchObject(expect.objectContaining(overrideOptions));
    });
  });
});
