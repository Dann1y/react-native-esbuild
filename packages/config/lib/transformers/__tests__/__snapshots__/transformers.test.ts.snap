// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBabelOptions should match snapshot 1`] = `
{
  "compact": false,
  "highlightCode": true,
  "minified": false,
  "root": "/root",
  "sourceMaps": false,
}
`;

exports[`getEsbuildOptions should match snapshot 1`] = `
{
  "bundle": true,
  "entryPoints": [
    "index.js",
  ],
  "format": "esm",
  "legalComments": "none",
  "logLevel": "silent",
  "metafile": false,
  "minify": false,
  "outfile": "main.bundlejs",
  "sourceRoot": ".",
  "sourcemap": true,
  "supported": {
    "for-of": false,
  },
  "target": "es6",
}
`;

exports[`getSwcOptions should match snapshot 1`] = `
{
  "inlineSourcesContent": false,
  "inputSourceMap": false,
  "isModule": true,
  "jsc": {
    "externalHelpers": true,
    "keepClassNames": true,
    "loose": false,
    "parser": {
      "exportDefaultFrom": true,
      "jsx": true,
      "syntax": "ecmascript",
    },
    "target": "es5",
  },
  "minify": false,
  "root": "/root",
  "sourceMaps": false,
}
`;
